### PR TITLE
Check gradient correctness in the UTs

### PR DESCRIPTION
**Pull request by SherlockNoMad**
Compare pytorch and ort computed gradient values in the unittests

### CODE REVIEW

File: orttraining/orttraining/python/training/ortmodule.py [hunk 1]
```diff
@@ -251,14 +251,18 @@ def forward(ctx, *inputs, **kwargs):
...
                 '''
 
                 # Use IO binding
                 # Push user output grads to ONNX backend.
                 backward_grad_output_ortvalue = []
-                for grad_output in grad_output[:len(self._onnx_graphs_info.backward_output_grad_names)]:
+                for grad_output in grad_outputs[:len(self._onnx_graphs_info.backward_output_grad_names)]:
+                    # Force torch tensors to be contiguous before converting into OrtValue
+                    if not grad_output.is_contiguous():
+                        grad_output = grad_output.contiguous()
```
Comment (iK1D) [inline orttraining/orttraining/python/training/ortmodule.py:264]: Looks like the variable scope in Python is different from other languages, this grad_output tensor is not released when going out of the for loop, otherwise, the data at grad_output.data_ptr() will be undefined when it goes to ORT.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 2]
```diff
@@ -370,143 +392,176 @@ def test_input_requires_grad_backward_creates_input_grad(device):
...
-        s.backward()
+        x = torch.randn(N, D_in, device=device, requires_grad=True)
+        run_step(pt_model1, pt_model2, x)
+        run_step(ort_model1, ort_model2, x)
 
-        assert x.grad is not None
-        for param in all_params:
-            assert param.grad is not None
-            param.grad = None
+        _check_model_gradients_match_and_reset(pt_model1, ort_model1)
```
Comment (iK1D) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:535]: When input x is requires_grad, do we need to copy it for out, and also use _assert_tensor_value_matches to check the grad tensors for input x and ort_x?

**Review by iK1D — APPROVED** [review]

File: orttraining/orttraining/python/training/ortmodule.py (hunk 1 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py:264]: Do you expect perf overhead with this? A possible memory call could take place here

Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/ortmodule.py:264]: @iK1D 
I think grad_output is taking a reference to grad_outputs, so the changes persist over the loop scope.

Comment (SherlockNoMad) [reply orttraining/orttraining/python/training/ortmodule.py:264]: @thiagocrepaldi 
Yes, there will be perf overhead if the tensor happens to be not contiguous, and a copy is inevitable, coz outvalue only support contiguous format. 

The hope is that pytorch doesn't have many cases where a non-contiguous tenser is returned. 
We caught this bug, because the gradient of .sum returns a sparse tensor.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py (hunk 2 above)
Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:535]: yes, need to clone the input data, if we wish to verify the gradient for inputs. 

see test_multiple_overlapping_forward_backward_calls, I have an example there.

**Comment by SherlockNoMad** [comment]
Yes, there will be perf overhead if the tensor happens to be not contiguous, and a copy is inevitable, coz outvalue only support contiguous format.
We caught this bug, because the gradient of .sum returns a sparse tensor.
The hope is that pytorch doesn't have many cases where a non-contiguous tenser is returned.

File: orttraining/orttraining/python/training/ortmodule.py (hunk 1 above)
Comment (thiagocrepaldi) [reply orttraining/orttraining/python/training/ortmodule.py:264]: Yeah, some copy is inevitable. I was just double checking whether we were on the same page :)
One example for such non contiguousness is transposing tensors, but there might be other cases out there.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 3]
```diff
@@ -140,6 +141,27 @@ def _get_bert_for_sequence_classification_sample_data(device):
...
+        pt_name, pt_param = pt_named_param
+        ort_name, ort_param = ort_named_param
+
+        assert pt_name in ort_name
+        _assert_tensor_value_matches(pt_param.grad, ort_param.grad)
+
+        if reset_grad:
+            pt_param.grad = None
+            ort_param.grad = None
+
```
[the remark this answers is not in the document]
Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:164]: thanks for the reminder. moved.

File: orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py [hunk 4]
```diff
@@ -370,143 +371,176 @@ def test_input_requires_grad_backward_creates_input_grad(device):
...
+        x1 = torch.randn(N, D_in, device=device)
+        run_step(pt_model0, pt_model1, x1)
+        run_step(ort_model0, ort_model1, x1)
+
+        _test_helpers.assert_gradients_match_and_reset_gradient(ort_model0, pt_model0, reset_gradient=False)
+        _test_helpers.assert_gradients_match_and_reset_gradient(ort_model1, pt_model1)
+
+        # Run task 2
+        x2 = torch.randn(N, D_in, device=device)
+        run_step(pt_model0, pt_model2, x1)
```
Comment (kit1980) [inline orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:490]: Should be `x2` here, not `x1`?

Comment (SherlockNoMad) [reply orttraining/orttraining/test/python/orttraining_test_ortmodule_api.py:490]: thanks for catching this. 
I will include this in my next PR

File: orttraining/orttraining/python/training/ortmodule.py (hunk 1 above)
Comment (iK1D) [reply orttraining/orttraining/python/training/ortmodule.py:264]: @SherlockNoMad If we don't call the contiguous(), grad_output is taking reference to grad_outputs, but when we call it, there should be a new tensor created inside the for loop. But it works. :-)
